src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,224 @@
+
+  private static final long   WORKER_TTL_MS          = 120_000L;
+  private static final long   PERIOD                 = 86400_000L;
+  private static final long   MAXIMUM_CHUNK_INTERVAL = 120_000L;
+  private static final long   DEFAULT_CHUNK_INTERVAL = 60_000L;
+  private static final long   MINIMUM_CHUNK_INTERVAL = 500L;
+  private static final int    CHUNK_SIZE             = 10000;
+  private static final double JITTER_MAX             = 0.20;
+
+  private final AccountsManager               accountsManager;
+  private final DirectoryReconciliationClient reconciliationClient;
+  private final DirectoryReconciliationCache  reconciliationCache;
+  private final String                        workerId;
+  private final SecureRandom                  random;
+
+  private ScheduledExecutorService executor;
+
+  public DirectoryReconciler(DirectoryReconciliationClient reconciliationClient,
+                             DirectoryReconciliationCache reconciliationCache,
+                             AccountsManager accountsManager) {
+    this.accountsManager      = accountsManager;
+    this.reconciliationClient = reconciliationClient;
+    this.reconciliationCache  = reconciliationCache;
+    this.random               = new SecureRandom();
+    this.workerId             = generateWorkerId(random);
+  }
+
+  private static String generateWorkerId(SecureRandom random) {
+    byte[] workerIdBytes = new byte[16];
+    random.nextBytes(workerIdBytes);
+    return Hex.toString(workerIdBytes);
+  }
+
+  @Override
+  public void start() {
+    start(Executors.newSingleThreadScheduledExecutor());
+  }
+
+  @VisibleForTesting
+  public void start(ScheduledExecutorService executor) {
+    this.executor = executor;
+
+    scheduleWithJitter(DEFAULT_CHUNK_INTERVAL);
+  }
+
+  @Override
+  public void stop() {
+    this.executor.shutdown();
+
+    try {
+      this.executor.awaitTermination(5000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException ex) {
+      throw new AssertionError(ex);
+    }
+  }
+
+  @Override
+  public void run() {
+    long scheduleDelayMs = DEFAULT_CHUNK_INTERVAL;
+
+    try {
+      long    intervalMs       = getBoundedChunkInterval(PERIOD * getAccountCount() / CHUNK_SIZE);
+      Instant nextIntervalTime = Instant.now().plus(intervalMs, ChronoUnit.MILLIS);
+
+      scheduleDelayMs = intervalMs;
+
+      if (reconciliationCache.claimActiveWork(workerId, WORKER_TTL_MS)) {
+        if (processChunk()) {
+          if (!reconciliationCache.isAccelerated()) {
+            scheduleDelayMs = getTimeUntilNextInterval(nextIntervalTime);
+            reconciliationCache.claimActiveWork(workerId, scheduleDelayMs);
+          } else {
+            scheduleDelayMs = MINIMUM_CHUNK_INTERVAL;
+          }
+        }
+      }
+    } catch (Exception ex) {
+      logger.warn("error in directory reconciliation", ex);
+    }
+
+    scheduleWithJitter(scheduleDelayMs);
+  }
```

Comment:
My experience with these kinds of long running tasks is that it can be better to have a single thread that loops, sleeps/waits in between intervals, and catches Throwable. Otherwise I have ended up with the thread disappearing and not noticing.